### DISCUSSION TITLE
Cached Samples not abiding by the Metropolis Local Rule.

### DISCUSSION THREAD

**Discussion by RPaul-Smith** [netket/netket#1685] [Q&A] [answered]
Hey everyone,

I am trying to use ```netket.experimental.TDVP``` to perform dynamics when quenching from the ground state of the tranverse field ising model to a non solvable point for transverse field ising model with a longitudinal field which leads to strongly chaotic dynamics.

To perform this I am using ```RBMSymm``` and the ```MetropolisLocal``` sampler to perform the monte carlo simulations for both the ground state search and the TDVP. However, while looking into this I just wanted to check the sampling to see if the sampler and variational state is returning valid transitions according to the local rule. It seems that the varational state doesn't return samples that abide by this local rule.

For example when set the ```seed=1000``` using this code

```python
import os
os.environ["JAX_PLATFORM_NAME"]="cpu"

import netket as nk
import numpy as np
from netket.operator.spin import sigmax,sigmaz
import netket.experimental as nkx
import flax.linen as nn
import sys
from tqdm import tqdm
import jax

N=6 #sets number of spins
hilb=nk.hilbert.Spin(s=1/2,N=N) #builds the Hilbert Space
graph=nk.graph.Chain(length=N, pbc=True)
samples=10 #sets the number of samples used in the VMC
LR=0.05 #sets the learning rate for the SGD
alpha=1 #sets the hidden unit density

model=nk.models.RBMSymm(alpha=alpha,param_dtype=np.complex128,symmetries=graph.translation_group()) #creates symmetric RBM
sampler = nk.sampler.MetropolisLocal(hilb,n_chains=1) #initialises the monte carlo sampler
vs = nk.vqs.MCState(sampler, model, n_samples=samples,seed=1000) #creates the initial variational state 
vs.init_parameters(nn.initializers.normal(stddev=0.01)) #initialises the parameters randomly

print(vs.samples)
```
then the set of samples it returns is 
```
[[[ 1.  1. -1. -1.  1. -1.]
  [-1.  1.  1. -1. -1.  1.]
  [-1.  1.  1.  1.  1.  1.]
  [ 1.  1.  1.  1. -1.  1.]
  [ 1. -1. -1. -1.  1.  1.]
  [-1. -1.  1. -1.  1.  1.]
  [ 1. -1. -1.  1.  1. -1.]
  [-1. -1. -1. -1.  1. -1.]
  [ 1.  1. -1. -1. -1.  1.]
  [-1.  1.  1. -1. -1.  1.]]]
```
which dont seem to abide by the metropolis local rule and this does happen with other seeds and when one isn't specified. Sorry if this is a very trivial question but is this a valid set of samples which do abide by the local rule or is there something I am misunderstaning with both the sampler or the variational state interface? Any help or advice would be greatly appreciated. Thank you!

**Comment by gcarleo** [comment, marked as answer]
Hi do you mean that the printed samples do not differ by a single spin flips? If so this is because samples are taken every sweep_size steps, where by default sweep_size is the number of degrees of freedom. This means there can be as many as N bit flips between one printed sample and the next one

**Reply by PhilipVinc** [reply]
@RPaul-Smith , in other words, if you simply specify
```python
sampler = nk.sampler.MetropolisLocal(hilb,n_chains=1, n_sweeps=1)
```

you will see that the condition is verified.

Do note that in the next release (and in nk master) `n_sweeps ` will be renamed to `sweep_size`.

Eg:
```python
   ...: model=nk.models.RBMSymm(alpha=alpha,param_dtype=np.complex128,symmetries=graph.translation_group()) #creates symmetric RBM
   ...: sampler = nk.sampler.MetropolisLocal(hilb,n_chains=1, sweep_size=1) #initialises the monte carlo sampler
   ...: vs = nk.vqs.MCState(sampler, model, n_samples=samples,seed=1000) #creates the initial variational state
   ...: vs.init_parameters(nn.initializers.normal(stddev=0.01)) #initialises the parameters randomly
   ...:
   ...: print(vs.samples)
[[[-1. -1.  1. -1. -1. -1.]
  [ 1. -1.  1. -1. -1. -1.]
  [ 1. -1.  1. -1.  1. -1.]
  [ 1. -1.  1.  1.  1. -1.]
  [ 1. -1.  1.  1.  1.  1.]
  [-1. -1.  1.  1.  1.  1.]
  [-1. -1.  1.  1.  1. -1.]
  [-1. -1.  1.  1.  1.  1.]
  [-1.  1.  1.  1.  1.  1.]
  [-1. -1.  1.  1.  1.  1.]]]
```
